### PR TITLE
Core: Fix possible proken in `Tasks.Builder.runSingleThreaded`

### LINKED ISSUE #10597: `o.a.i.util.Tasks.Builder.runSingleThreaded` possbily broken
### Apache Iceberg version

main (development)

### Query engine

None

### Please describe the bug 🐞

The evaluation at the end of `org.apache.iceberg.util.Tasks.Builder#runSingleThreaded` is broken, because `threw` can only be `false` at that point.

### PR DESCRIPTION
**Pull request by tlm365**
Closes #10597.

### CODE REVIEW

**Review by snazy — COMMENTED** [review]
Wonder if a few tests would be good for this rather central piece of code?

File: core/src/main/java/org/apache/iceberg/util/Tasks.java [hunk 1]
```diff
@@ -226,13 +226,11 @@ private <E extends Exception> boolean runSingleThreaded(
             }
 
             if (stopOnFailure) {
+              threw = true;
```
Comment (amogh-jahagirdar) [inline core/src/main/java/org/apache/iceberg/util/Tasks.java:229]: I don't think this the right place to set this? Shouldn't `threw` be set for all cases where we catch an execption, not just the `stopOnFailure` case?

**Review by amogh-jahagirdar — CHANGES_REQUESTED** [review]
@snazy @tlm365 Thanks for bringing this up but I read the code further and I don't think it's broken. It looks like the intention of the `threw` variable is to handle cases which are not handled by the catch block (some arbitrary throwable that's not an `Exception`).  
Perhaps we should rename the variable to be something a bit more clear. I also think the real simplification is to always just return true on line 281 since at that point we know that `threw` is false.

If we're able to provide a `Tasks.run` call which does not return the expected boolean result, that'd help show if there's a legitimate issue here or not. On a side note, I think in the long run, we may want to consider using [Failsafe](https://failsafe.dev/) to help simplify the Tasks implementation, but that's of course a long run thing imo.

**Comment by tlm365** [comment]
> ... It looks like the intention of the `threw` variable is to handle cases which are not handled by the catch block (**some arbitrary throwable that's not an `Exception`**).

@amogh-jahagirdar thanks so much for reviewing, but I don't think so. Inner the `try` block, the [runTaskWithRetry](https://github.com/apache/iceberg/blob/d3cb1b696a1631a9cca4619f93252cd0a985fbfc/core/src/main/java/org/apache/iceberg/util/Tasks.java#L403) function and `succeeded.add(item)` only throw `Exception`'s extenders (if errors). Then, `threw` is always `false`?

File: core/src/main/java/org/apache/iceberg/util/Tasks.java (hunk 1 above)
Comment (tlm365) [reply core/src/main/java/org/apache/iceberg/util/Tasks.java:229]: Oh, thank you, nice catch. I agree, place `threw` within catch clause is more exact.

**Pull request closed without merging**